File: samples/leetcode/two_pointers/valid_palindrome.cpp
#include <string>
#include <cassert>

using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        size_t j = s.size() -1;
        for (size_t i = 0; i < s.size()/2; i++) {
            if (s[i] != s[j]) {
                return false;
            }
            --j;
        }
        return true;
    }
};

int main() {
    auto solution = Solution();

    assert(solution.isPalindrome("amanaplanacanalpanama"));
    assert(!solution.isPalindrome("race a car"));

    return 0;
}

File: bazel/samples/leetcode/two_pointers/valid_palindrome.rs
use std::error::Error;

struct Solution;

impl Solution {
    pub fn is_palindrome(s: String) -> bool {
        let chars: Vec<char> = s.chars().collect();
        let mut j = chars.len().saturating_sub(1); // Prevent underflow
        for i in 0..chars.len() / 2 { // Only need to iterate through half
            let a = chars[i];
            let b = chars[j];

            if a != b {
                return false;
            }

            if j == 0 {
                break; // Prevent underflow
            }
            j -= 1;
        }
        true
    }
}

fn main() -> Result<(), Box<dyn Error>> {
    assert!(Solution::is_palindrome("amanaplanacanalpanama".to_string()));
    assert!(!Solution::is_palindrome("race a car".to_string()));
    Ok(())
}